## Практическое домашнее задание на проверку знаний по модулю 1.


##### Тема тестирования - базовый эксплоративный анализ и аналитическое мышление. Вам будет предоставлен датасет и 25 вопросов касаемо данного датасета. Ваша задача - подумать как найти ответы на поставленные вопросы ниже. За каждый правильно оформленный и предоставленный ответ начисляется *4 балла*.

#### Ответ на решение присылать ссылкой на ваш гитхаб репозиторий на почту maratmovlamov2017@gmail.com

#### Сроки выполнения практического домашнего задания - ровно 7 календарных дней (168 часов) с момента публикации в группе. Прошу приступать к решению задач порционно. Прошу повторить дополнительные материалы, предоставленные в качестве самостоятельного изучения в конце презентаций, они вам понадобятся.



#### Критерии оценки присланного решения:

* Правильность полученного ответа - на каждый поставленный вопрос есть один и только один правильный ответ. В случае неправильного ответа на поставленный вопрос, предоставленный ответ на него считается проваленным, даже если код написан правильно и красиво. Общее количество баллов, которые можно получить ответив на все вопросы верно - 25 * 4 = 100 баллов, которые шкалируются к 10-ти бальной шкале.

___

* Аккуратность, правильность и читаемость кода - старайтесь писать код без повторного использования методов и ненужных переменных (помните про принцип DRY - don't repeat yourself). Плюсом будет, если ваш код будет упакован в лаконичную и красиво написанную функцию. Если вся работа выполненна аккуратно, без спагетти кода, к общему результату работы начисляется 1 балл.

---

* На предоставленные вопросы, дописывается комментарий для проверки вашего мышления и поиска ответа в виде Markdown комментария. В случае если комментарий на поставленный вопрос не предоставлен снимается 0.1 балл за каждый такой вопрос. То есть общий штраф за отсутсвие комментариев по всем вопросам 25 * -0.1 =  -2.5 балла.


#### Удачи и да прибудет с Вами сила :)



In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("./dataset/Spotify_Youtube.csv")

In [45]:
df.head()

Unnamed: 0    Artist                                        Url_spotify   
0           0  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...  \
1           1  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
2           2  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
3           3  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
4           4  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   

                                           Track   
0                                 Feel Good Inc.  \
1                                Rhinestone Eyes   
2  New Gold (feat. Tame Impala and Bootie Brown)   
3                             On Melancholy Hill   
4                                 Clint Eastwood   

                                           Album Album_type   
0                                     Demon Days      album  \
1                                  Plastic Beach      album   
2  New Gold (feat. Tame Impala and Bootie Brown)     single   
3                                  Plastic Beach      album   
4                                       Gorillaz      album   

                                    Uri  Danceability  Energy   Key  ...   
0  spotify:track:0d28khcov6AiegSCpG5TuT         0.818   0.705   6.0  ...  \
1  spotify:track:1foMv2HQwfQ2vntFf9HFeG         0.676   0.703   8.0  ...   
2  spotify:track:64dLd6rVqDLtkXFYrEUHIU         0.695   0.923   1.0  ...   
3  spotify:track:0q6LuUqGLUiCPP1cbdwFs3         0.689   0.739   2.0  ...   
4  spotify:track:7yMiX7n9SBvadzox8T5jzT         0.663   0.694  10.0  ...   

                                   Url_youtube   
0  https://www.youtube.com/watch?v=HyHNuVaZJ-k  \
1  https://www.youtube.com/watch?v=yYDmaexVHic   
2  https://www.youtube.com/watch?v=qJa-VFwPpYA   
3  https://www.youtube.com/watch?v=04mfKJWDSzI   
4  https://www.youtube.com/watch?v=1V_xRb0x9aw   

                                               Title   Channel        Views   
0         Gorillaz - Feel Good Inc. (Official Video)  Gorillaz  693555221.0  \
1  Gorillaz - Rhinestone Eyes [Storyboard Film] (...  Gorillaz   72011645.0   
2  Gorillaz - New Gold ft. Tame Impala & Bootie B...  Gorillaz    8435055.0   
3     Gorillaz - On Melancholy Hill (Official Video)  Gorillaz  211754952.0   
4         Gorillaz - Clint Eastwood (Official Video)  Gorillaz  618480958.0   

       Likes  Comments                                        Description   
0  6220896.0  169907.0  Official HD Video for Gorillaz' fantastic trac...  \
1  1079128.0   31003.0  The official video for Gorillaz - Rhinestone E...   
2   282142.0    7399.0  Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3  1788577.0   55229.0  Follow Gorillaz online:\nhttp://gorillaz.com \...   
4  6197318.0  155930.0  The official music video for Gorillaz - Clint ...   

   Licensed official_video        Stream  
0      True           True  1.040235e+09  
1      True           True  3.100837e+08  
2      True           True  6.306347e+07  
3      True           True  4.346636e+08  
4      True           True  6.172597e+08  

[5 rows x 28 columns]

### Структура датасета:

----

26 признаков, описание которых следует ниже:



- **Track**: name of the song, as visible on the Spotify platform.
- **Artist**: name of the artist.
- **Url_spotify**: the Url of the artist.
- **Album**: the album in wich the song is contained on Spotify.
- **Album_type**: indicates if the song is relesead on Spotify as a single or contained in an album.
- **Uri**: a spotify link used to find the song through the API.
- **Danceability**: describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.
- **Energy:** is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.
- **Key**: the key the track is in. Integers map to pitches using standard Pitch Class notation. E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.
- **Loudness**: the overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typically range between -60 and 0 db.
- **Speechiness**: detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.
- **Acousticness**: a confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.
- **Instrumentalness**: predicts whether a track contains no vocals. "Ooh" and "aah" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly "vocal". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.
- **Liveness**: detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.
- **Valence**: a measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
- **Tempo**: the overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.
- **Duration_ms**: the duration of the track in milliseconds.
- **Stream**: number of streams of the song on Spotify.
- **Url_youtube**: url of the video linked to the song on Youtube, if it have any.
- **Title**: title of the videoclip on youtube.
- **Channel**: name of the channel that have published the video.
- **Views**: number of views.
- **Likes**: number of likes.
- **Comments**: number of comments.
- **Description**: description of the video on Youtube.
- **Licensed**: Indicates whether the video represents licensed content, which means that the content was uploaded to a channel linked to a YouTube content partner and then claimed by that partner. official_video: boolean value that indicates if the video found is the official video of the song.

### Вопросы:

#### 1. Какая песня является наиболее транслируемой в датасете (признак Stream)?

Необходимо найти максимальное значение в колонке Stream

In [92]:
#    через idxmax ищем максимальное по колонке Stream, выводим значение по индексу
df.loc[df['Stream'].idxmax()][['Track', 'Artist']]

Track     Blinding Lights
Artist         The Weeknd
Name: 15250, dtype: object

Ответ: Наиболее транслируемой песней в Spotify является песня: 

"Blinding Lights"- The Weeknd

#### 2. У какого исполнителя больше всего просмотров на YouTube?

Нужно сгруппировать по исполнителям - так мы получим уникальных исполнителей.
Затем аггрегировать по значениям Views - складываем все просмотры видео, у которых один и тот же исполнитель.
И по максимальному Views выводим исполнителя.

In [89]:
#    группируем фрейм по исполнителям, через agg считаем сумму просмотров на youtube
df_gp_artist = df.groupby('Artist').agg({'Views': 'sum'})

#    ищем максимальное по Views
df_gp_artist['Views'].idxmax()

'Ed Sheeran'

Ответ: Больше всего просмотров на YouTube имеет:

Ed Sheeran

#### 3. Какова средняя продолжительность всех песен в минутах?

Необходимо найти среднее по колонке Duration_ms, затем перевести в минуты

In [97]:
#    ищем среднее по колонке Duration_ms, переводим в минуты
mean_dur = df['Duration_ms'].mean() / 1000 / 60

print(f'Средняя продолжительноcть песен:{mean_dur: .4f} минут')

Средняя продолжительноcть песен: 3.7453 минут


Ответ: средння продолжительность песен - примерно 3.75 минут или 3 минуты и 45 секунд

#### 4. Какая нота (Key) чаще всего используется в песнях?

Самое частое для значение тональности можно найти через моду по колонке Key.

In [106]:
#    ищем моду в Key
df['Key'].mode()

0    0.0
Name: Key, dtype: float64

Ответ: Самая частая нота в песнях - 0, она же "До", она же "C"

#### 5. Какая самая популярная песня на YouTube?

Вероятно, имеется ввиду песня с бóльшим числом просмотром. В таком случае ищем максимальное значение в Views

In [107]:
#    ищем максимальное Views
df.loc[df['Views'].idxmax()][['Track', 'Artist']]

Track      Despacito
Artist    Luis Fonsi
Name: 1147, dtype: object

Ответ: Самая популярная песня на YouTube:

"Despacito" - Luis Fonsi

#### 6. У какой песни самый высокий балл за танцевальность (danceability )?

В первую очередь нам нужно найти максимальное значение для Danceability.
Затем искать по этому значению треки, которые ему соответствуют.

In [104]:
#    самый высокий балл может быть не у одной песни, 
#    поэтому сначала найдем максимальное значение "танцевальности", а затем найдем треки по этому значению
max_dance = df.loc[df['Danceability'].idxmax()]['Danceability']
df.loc[df['Danceability'] == max_dance][['Track', 'Artist']]

Track     Artist
750       Give It To Me  Timbaland
16555  CAN'T TOUCH THIS        BIA
18822      Funky Friday       Dave

Ответ: песни с максимальной танцевальностью:

- "Give It To Me" -	Timbaland

- "CAN'T TOUCH THIS" - BIA

- "Funky Friday" - Dave

#### 7. У какой песни самый высокий показатель энергии (energy score)?

Также ищем максимальное значение в Energy, а затем по этому значению ищем строки, содержащие его.

In [106]:
#   аналогично предыдущемму заданию
max_energy = df.loc[df['Energy'].idxmax()]['Energy']
df.loc[df['Energy'] == max_energy][['Track', 'Artist']]

Track              Artist
20581   Rain and Thunderstorm, Pt. 3  Rain Fruits Sounds
20582   Rain and Thunderstorm, Pt. 4  Rain Fruits Sounds
20585   Rain and Thunderstorm, Pt. 7  Rain Fruits Sounds
20586   Rain and Thunderstorm, Pt. 6  Rain Fruits Sounds
20587  Rain and Thunderstorm, Pt. 33  Rain Fruits Sounds
20652          Gentle Piano Melodies      Sleepy Penguin

Ответ: песни с максимальной энергичностью:
- "Rain and Thunderstorm, Pt. 3" - Rain Fruits Sounds
- "Rain and Thunderstorm, Pt. 4" - Rain Fruits Sounds
- "Rain and Thunderstorm, Pt. 7" - Rain Fruits Sounds
- "Rain and Thunderstorm, Pt. 6" - Rain Fruits Sounds
- "Rain and Thunderstorm, Pt. 33" - Rain Fruits Sounds
- "Gentle Piano Melodies" - Sleepy Penguin

#### 8. Какая песня имеет наивысший балл valence (valence score)?

Ищем максимальное Valence, ищем треки с максимальным Valence. 

In [108]:
#    аналогично двум предыдущим
max_valence = df.loc[df['Valence'].idxmax()]['Valence']
df.loc[df['Valence'] == max_valence][['Track', 'Artist']]

Track    Artist
18514  Spider Dance  Toby Fox

Ответ: песня с максимальным баллом Valence: 

"Spider Dance" - Toby Fox

#### 9. Будут ли песни с более высокой «valence» (музыкальной позитивностью) более популярны как на Spotify, так и на YouTube? Чтобы ответить на этот вопрос, подсчитайте среднюю оценку «valence» для песен в верхнем квартиле «Streams» и «Views» и сравните ее с общей средней «valence».

В принципе, алгоритм действий описан в вопросе.

In [111]:
#    создаем фрейм со значениями Valence выше верхнего квартиля
df_75 = df.loc[df['Valence'] > df['Valence'].quantile(0.75)]

#    вычисляем разницу по Stream и Views для общего среднего и значений выше верхнего квартиля (в процентах)
diff_stream = df_75['Stream'].mean() / df['Stream'].mean() * 100 - 100
diff_views = df_75['Views'].mean() / df['Views'].mean() * 100 - 100

print(f'Разница для YouTube {diff_views: .4f}%')
print(f'Разница для Spotify {diff_stream: .4f}%')

Разница для YouTube  13.3521%
Разница для Spotify -7.5233%


Ответ: Видим, что в среднем для песен с высоким Valence количество просмотром выше примерно на 13.35%.
Однако для Spotify песни с высоким Valence в среднем набирают меньше прослушиваний на 7.52%

#### 10. Какой канал на YouTube имеет наибольшее количество просмотров?

Какие-то видео из датасета могут быть с одного канала (внезапно). Поэтому нам нужно сгруппировать таблицу по признаку Channel и сложить просмотры. После этого сможем найти максимальное.

In [114]:
#   группируем по Channel и суммируем все Views
df_gp_channel = df.groupby('Channel').agg({'Views': 'sum'})

#   ищем максимальное число просмотров в полученном фрейме
df_gp_channel['Views'].idxmax()

'T-Series'

Ответ: Канал с наибольшим количеством просмотров на YouTube имеет: 

"T-Series"

#### 11. У песни какого исполнителя больше всего комментариев на YouTube?

Нужно найти максимальное в колонке Comments и определить исполнителя.

In [116]:
#    ищем максимальное по Comments
df.loc[df['Comments'].idxmax()][['Track', 'Artist']]

Track     Dynamite
Artist         BTS
Name: 14561, dtype: object

Ответ: Больше всего комментариев на YouTube у песни: 

"Dynamite" - BTS

#### 12. Какой самый распространенный темп (tempo) песен?

Напрашивается сразу использовать mode на колонку Tempo. Но так как значения в Tempo вещественные, мы не получим внятный результат - необходимо округлить значения Tempo, например до целой части.

Также нам потребуется избавиться от NaN'ов в колонке - round не умеет округлять NaN'ы.

In [122]:
#    округляем Tempo до целой части, проверяем на NaN'ы
df_round_tempo = df['Tempo']
df_round_tempo.isnull().sum()


2

In [121]:
#    избавляемся от NaN'ов, округляем Tempo и ищем моду
df_round_tempo.dropna(inplace=True)
df_round_tempo.apply(lambda x: round(x)).mode()[0]

120

Ответ: при округлугленных значениях темпа до целой части самым распространенным является: 120 bpm

#### 13. Какая песня имеет самый высокий показатель по speechiness?

Ищем максимальное Speechiness, по нему ищем треки.

In [123]:
#    находим максимальное значение Speechiness и ищем по этому значению строки
max_speechiness = df.loc[df['Speechiness'].idxmax()]['Speechiness']
df.loc[df['Speechiness'] == max_speechiness][['Track', 'Artist']]

Track   
5158  Teil 7 - Sherlock Holmes und der blinde Bettle...  \

                      Artist  
5158  Sir Arthur Conan Doyle

In [124]:
#    глянем полное название
df.iloc[5158]['Track']

'Teil 7 - Sherlock Holmes und der blinde Bettler - Die neuen Abenteuer, Folge 50'

Ответ: Самый высокий показатель по Speechiness имеет трек: 

"Teil 7 - Sherlock Holmes und der blinde Bettler - Die neuen Abenteuer, Folge 50" - Sir Arthur Conan Doyle

#### 14. Какая самая громкая песня в наборе данных?

Ищем максимальное в Loudness, по нему ищем треки.

In [33]:
#    находим максимальное значение Loudness и ищем по этому значению строки
max_loudness = df.loc[df['Loudness'].idxmax()]['Loudness']
df.loc[df['Loudness'] == max_loudness][['Track', 'Artist']]

Track     Artist
9423  Siren  Sub Focus

Ответ: Самая громкая песня:

"Siren" - Sub Focus

#### 15. У какого артиста самый высокий средний балл danceability?

Нужно сгруппировать по исполнителям и найти среднее их Danceability. Далее ищем максимальное и на всякий случай ищем по максимальному треки (вдруг совпадут средние по Danceability).

In [50]:
#    группируем по исполнителям, с помощью agg ищем среднее всех Danceability треков артиста
df_gp_artist = df.groupby('Artist').agg({'Danceability': 'mean'})

#    на всякий случай поищем максимальное Danceability среди других исполнителей
max_dance = df_gp_artist.loc[df_gp_artist['Danceability'].idxmax()]['Danceability']
df_gp_artist.loc[df_gp_artist['Danceability'] == max_dance]

Danceability
Artist                
Saweetie        0.8858

Ответ: Самый высокий балл Danceabillity и исполнителя:

Saweetie

#### 16. Как средний показатель «speechiness» песен зависит от «Album_type»? Песни, выпущенные как синглы, чаще содержат больше слов согласно анализу?

Нужно вычислить отдельно среднюю Speechiness для синглов и среднюю Speechiness для треков в составе альбома. Сравнить.

In [51]:
#    отберем в отдельную таблицу треки-синглы и вычислим среднее по их Speechiness
df_single = df.loc[df['Album_type'] == 'single']
mean_s = df_single['Speechiness'].mean()

#    отберем в отдельную таблицу треки в составе альбомов и вычислим среднее по их Speechiness
df_album = df.loc[df['Album_type'] == 'album']
mean_a = df_album['Speechiness'].mean()

#    посчитаем разницу в процентах
print(f'Разница между средними для синглов и альбомов: {mean_s / mean_a * 100 - 100: .4f}%')

Разница между средними для синглов и альбомов:  9.4490%


Ответ: Speechiness в синглах в среднем выше на 9.45% по сравнению с песнями в составе альбомов

#### 17. Каково среднее количество прослушиваний песен, которые входят в топ 10% лучших по метрике танцевальности?

Можно сортировать таблицу по Danceability по убыванию. Затем взять первых 10% строк - так мы получим ТОП 10% лучших по танцевальности. 

И далее посчитать среднюю по Stream в полученной таблице.

In [55]:
#    создадим отдельную таблицу с отсортированными Danceability в порядке убывания 
df_sorted = df.sort_values(by='Danceability', ascending=False)

#    отберем 10% самых танцевальных
df_10_dance = df_sorted[:int(round(df.shape[0] / 10))]

#    посчитаем среднее прослушиваний
top_10_dance = df_10_dance['Stream'].mean()
print(f'Среднее количество прослушиваний по TOP-10% танцевальности:{top_10_dance: .2f}')

Среднее количество прослушиваний по TOP-10% танцевальности: 151903593.13


Ответ: TOP-10% песен по высокой танцевальности прослушивают в среднем почти 152 млн. раз

#### 18. Для песен, которые входят в топ 10% лучших с точки зрения «danceability» метрики, как соотносится их средний показатель «tempo» со средним показателем «tempo» всех песен?

Таблица ТОП-10% у нас уже есть. Нам остается только найти соотношение средних показателей Tempo общей танцевальности и высокой.

In [72]:
#    таблица df_10_dance уже создана в предыдущем задании, воспользуемся ей

#    найдем разницу средних общего Tempo и Tempo в TOP-10% Danceability
diff_tempo = (df_10_dance['Tempo'].mean() / df['Tempo'].mean()) * 100 - 100

#    выводим результаты
print(f"Средний темп TOP-10% танцевальности:{df_10_dance['Tempo'].mean(): .2f} bpm")
print(f"Средний темп общий:{df['Tempo'].mean(): .2f} bpm")
print(f'Разница средних между темпом TOP-10% танцевальности и общим:{diff_tempo: .2f}%')

Средний темп TOP-10% танцевальности: 115.07 bpm
Средний темп общий: 120.64 bpm
Разница средних между темпом TOP-10% танцевальности и общим:-4.62%


Ответ: Темп песен с высокой танцевальностью в среднем ниже на 4.62% по сравнению со средним темпом всех песен

#### 19. Какова корреляция между «valence» метрикой песни и ее популярностью на YouTube (по показателю «Views») и Spotify (по показателю «Stream»)? Счастливые песни более популярны? Valency отвечает за жизнерадостность песни.

Тут просто используем метод corr для вычисления показателя корреляции между двумя колонками.

In [34]:
#    воспользуемся corr для нахождения корреляции
corr_valence_views = df['Valence']. corr(df['Views'])
corr_valence_stream = df['Valence']. corr(df['Stream'])

print(f'Показатель корреляции между Valence и Views: {corr_valence_views:.4f}')
print(f'Показатель корреляции между Valence и Stream: {corr_valence_stream:.4f}')

Показатель корреляции между Valence и Views: 0.0354
Показатель корреляции между Valence и Stream: -0.0121


Ответ: Корреляция Valence c Views и Stream имеет показатели 0.0354 и -0.0121 соответственно. Предполагаю, что жизнерадостность песни не влияет на ее популярность.

#### 20. Какова средняя громкость песен, которые входят в топ 10% лучших по энергетике (Energy)?

Создаем ТОП-10% лучших по Energy и в полученной таблице ищем среднее по Loudness.

In [79]:
#    создадим таблицу с отсортированными в порядке убывания значениями Energy
df_energy = df.sort_values(by='Energy', ascending=False)

#   отберем 10% самых энергичных
df_10_energy = df_sorted[:int(round(df.shape[0] / 10))]

#    посчитаем среднюю громкость от этих 10% энергичных
mean_loudness_10 = df_10_energy['Loudness'].mean()
print(f'Средняя громкость TOP-10% по энергичности:{mean_loudness_10:.5f} dB')

Средняя громкость TOP-10% по энергичности:-6.56626 dB


Ответ: Средняя громкость 10% лучших по Energy песен: -6.56 dB

#### 21. У какого исполнителя больше всего песен, которые входят в топ-10% как по просмотрам на YouTube, так и по стримам Spotify?

Создадим две таблицы: ТОП-10% по Views и ТОП-10% по Stream. Затем соединяем их через merge по признаку 'Unnamed: 0'. Значения в 'Unnamed: 0' уникальны, вероятно используется в качестве ID. Объединение таблиц будет по принципу inner, то есть в объединенной таблице будут только те строки, которые есть в каждой из объединяемых таблицах. То есть, у нас будут песни, которые входят в топ-10% как по просмотрам на YouTube, так и по стримам Spotify.

Посчитать количество вхождений исполнителей можно через value_counts.

In [91]:
#    создадим таблицу с отсортированными по убыванию Views
df_views = df.sort_values(by='Views', ascending=False)

#    отберем 10% самых просматриваемых
df_10_views = df_views[:int(round(df.shape[0] / 10))]
df_10_views['Views'].head()

1147     8.079649e+09
365      8.079647e+09
12452    5.908398e+09
14580    5.773798e+09
12469    5.773797e+09
Name: Views, dtype: float64

In [92]:
#    создадим таблицу с отсортированными по убыванию Stream
df_stream = df.sort_values(by='Stream', ascending=False)

#    отберем 10% самых прослушиваемых
df_10_stream = df_stream[:int(round(df.shape[0] / 10))]
df_10_stream['Stream'].head()

15250    3.386520e+09
12452    3.362005e+09
19186    2.634013e+09
17937    2.594927e+09
17445    2.538330e+09
Name: Stream, dtype: float64

In [94]:
#    соединим наши таблицы так, чтобы в нее попали уникальные строки (inner join), 
#    таким образом, получим таблицу с песнями, которые в топе и youtube, и spotify
df_10_views_stream = df_10_views.merge(df_10_stream, on='Unnamed: 0')
df_10_views_stream

Unnamed: 0       Artist_x   
0           1147     Luis Fonsi  \
1            365   Daddy Yankee   
2          12452     Ed Sheeran   
3          14580   Charlie Puth   
4          12469    Wiz Khalifa   
...          ...            ...   
1090       12038       Paramore   
1091       18809    Niall Horan   
1092       13700     The Script   
1093       13419    Soundgarden   
1094       14070  Justin Bieber   

                                          Url_spotify_x   
0     https://open.spotify.com/artist/4V8Sr092TqfHkf...  \
1     https://open.spotify.com/artist/4VMYDCV2IEDYJA...   
2     https://open.spotify.com/artist/6eUKZXaKkcviH0...   
3     https://open.spotify.com/artist/6VuMaDnrHyPL1p...   
4     https://open.spotify.com/artist/137W8MRPWKqSmr...   
...                                                 ...   
1090  https://open.spotify.com/artist/74XFHRwlV6OrjE...   
1091  https://open.spotify.com/artist/1Hsdzj7Dlq2I7t...   
1092  https://open.spotify.com/artist/3AQRLZ9PuTAozP...   
1093  https://open.spotify.com/artist/5xUf6j4upBrXZP...   
1094  https://open.spotify.com/artist/1uNFoZAHBGtllm...   

                                 Track_x                             Album_x   
0                              Despacito                                VIDA  \
1                              Despacito                                VIDA   
2                           Shape of You                          ÷ (Deluxe)   
3     See You Again (feat. Charlie Puth)  See You Again (feat. Charlie Puth)   
4     See You Again (feat. Charlie Puth)  See You Again (feat. Charlie Puth)   
...                                  ...                                 ...   
1090                      Still into You                            Paramore   
1091                          Slow Hands                    Flicker (Deluxe)   
1092                         Superheroes            No Sound Without Silence   
1093                      Black Hole Sun       Superunknown (Deluxe Edition)   
1094                               Ghost                             Justice   

     Album_type_x                                 Uri_x  Danceability_x   
0           album  spotify:track:6habFhsOp2NvshLv26DqMb           0.655  \
1           album  spotify:track:6habFhsOp2NvshLv26DqMb           0.655   
2           album  spotify:track:7qiZfU4dY1lWllzX7mPBI3           0.825   
3          single  spotify:track:2JzZzZUQj3Qff7wapcbKjc           0.689   
4          single  spotify:track:2JzZzZUQj3Qff7wapcbKjc           0.689   
...           ...                                   ...             ...   
1090        album  spotify:track:1yjY7rpaAQvKwpdUliHx0d           0.602   
1091        album  spotify:track:2cOUlop2ZakjOKe4WinesN           0.735   
1092        album  spotify:track:37sINbJZcFdHFAsVNsPq1i           0.490   
1093        album  spotify:track:2EoOZnxNgtmZaD8uUmz2nD           0.350   
1094        album  spotify:track:6I3mqTwhRpn34SLVafSH7G           0.601   

      Energy_x  Key_x  ...                                Url_youtube_y   
0        0.797    2.0  ...  https://www.youtube.com/watch?v=kJQP7kiw5Fk  \
1        0.797    2.0  ...  https://www.youtube.com/watch?v=kJQP7kiw5Fk   
2        0.652    1.0  ...  https://www.youtube.com/watch?v=JGwWNGJdvx8   
3        0.481   10.0  ...  https://www.youtube.com/watch?v=RgKAFK5djSk   
4        0.481   10.0  ...  https://www.youtube.com/watch?v=RgKAFK5djSk   
...        ...    ...  ...                                          ...   
1090     0.923    5.0  ...  https://www.youtube.com/watch?v=OblL026SvD4   
1091     0.425    0.0  ...  https://www.youtube.com/watch?v=nBmNcLBaPUE   
1092     0.885    3.0  ...  https://www.youtube.com/watch?v=WIm1GgfRz6M   
1093     0.828    6.0  ...  https://www.youtube.com/watch?v=3mbBbFH9fAg   
1094     0.741    2.0  ...  https://www.youtube.com/watch?v=Fp8msa5uYsc   

                                                Title_y          Channel_y   
0               Luis Fonsi - Despac

In [95]:
#    теперь в полученной таблице через value_counts посчитаем уникальных исполнителей и их количество
df_10_views_stream['Artist_x'].value_counts()

Artist_x
Calvin Harris      9
Dua Lipa           9
Coldplay           9
Ariana Grande      9
Imagine Dragons    8
                  ..
Kodak Black        1
AURORA             1
The Killers        1
Vance Joy          1
Soundgarden        1
Name: count, Length: 480, dtype: int64

Ответ: Больше всего песен, которые входят в топ-10% как по просмотрам на YouTube, так и по стримам Spotify, имеет:

- Calvin Harris      
- Dua Lipa           
- Coldplay           
- Ariana Grande      

#### 22. Какова корреляция между «energy» песни и ее «loudness»?

Используем corr.

In [63]:
#    пользуемся методом corr для значений Energy и Loudness
df['Energy'].corr(df['Loudness'])

0.7448445555528459

Ответ: Коррелиция между Energy и Loudness сильная. Вероятно, связано с тем, что при расчете метрики Energy используется метрика Loudness. Также в описании метрики Energy видим: "Perceptual features contributing to this attribute include dynamic range, perceived *loudness*, timbre, onset rate, and general entropy."

#### 23. Каково среднее количество лайков на YouTube для песен, которые находятся в верхнем квартиле (25%) и как это соотносится с общим средним показателем?

Не совсем понятно, про верхний квартиль какого признака идет речь, поэтому я решил пройти по всем возможным признакам и сделать какие-то выводы.

Нужно сформировать таблицу, в котором будут лежат значения больше верхнего квартиля признака.
Затем сравнить средние лайков в этой таблице и лайков исходной таблицы.

In [103]:
#    лист с числовыми признаками
list_features = ['Danceability',
                 'Energy',
                 'Loudness',
                 'Speechiness',
                 'Acousticness',
                 'Instrumentalness',
                 'Liveness',
                 'Valence',
                 'Tempo',
                 'Duration_ms',
                 'Views',
                 'Likes',
                 'Comments',
                 'Stream']

#    в цикле:
#    формируем таблицу с верхним квартилем по признаку,
#    вычисляем разницу в процентах между средними верхнеквартильным и общим
for feature in list_features:
    df_75 = df.loc[df[feature] > df[feature].quantile(0.75)]
    diff = df_75['Likes'].mean() / df['Likes'].mean() * 100 - 100
    print(f'Разница для Likes в верхнем квартиле {feature}:{diff: .2f}%')

Разница для Likes в верхнем квартиле Danceability: 33.75%
Разница для Likes в верхнем квартиле Energy: 3.74%
Разница для Likes в верхнем квартиле Loudness: 47.20%
Разница для Likes в верхнем квартиле Speechiness: 12.90%
Разница для Likes в верхнем квартиле Acousticness:-34.37%
Разница для Likes в верхнем квартиле Instrumentalness:-37.75%
Разница для Likes в верхнем квартиле Liveness:-7.11%
Разница для Likes в верхнем квартиле Valence:-0.75%
Разница для Likes в верхнем квартиле Tempo:-3.57%
Разница для Likes в верхнем квартиле Duration_ms:-9.26%
Разница для Likes в верхнем квартиле Views: 237.55%
Разница для Likes в верхнем квартиле Likes: 247.89%
Разница для Likes в верхнем квартиле Comments: 239.34%
Разница для Likes в верхнем квартиле Stream: 182.34%


Ответ: Из полученных расчетов наверное самые интересные - Views, Likes, Comments. Среднее по Likes в верхнем квантиле для Views, Likes, Comments больше общего среднего по Likes на 237.55%, 247.89% и 239.34% соответственно

#### 24. Есть ли существенная разница в средней «valence» метрикой для песен, выпущенных как синглы, по сравнению с песнями, выпущенными в виде альбома?

Ищем средние Valence для песен типа сингл и альбомных, сравниваем эти значения.

In [78]:
#    через loc выбираем строки по условию single или album, считаем среднее
mean_val_single = df.loc[df['Album_type'] == 'single']['Valence'].mean()
mean_val_album = df.loc[df['Album_type'] == 'album']['Valence'].mean()

print(f'Разница Valence между Single и Album:{mean_val_single / mean_val_album * 100 - 100: .5f}%')

Разница Valence между Single и Album: 0.06702%


Ответ: Среднее Valence в синглах и среднее Valence в составе альбома отличаются несущестенно. Разница составляет примерно 0.067%

#### 25. Каково среднее количество «комментариев» на YouTube для песен с показателем «speechiness» выше 0,66 (что указывает на большое количество произнесенных слов) и как это соотносится с общим средним показателем?

Формируем таблицу со значениями Speechiness > 0.66, ищем в ней среднее по Comments. Сравниваем это среднее с общим средним Comments.

In [105]:
#    ищем общую среднюю по Comments и среднню по Comments с высокой Speechiness
mean_comments = df['Comments'].mean()
mean_comments_high_speech = df.loc[df['Speechiness'] > 0.66]['Comments'].mean()

print(f'Разница Comments между высокой Speechiness и общей:{mean_comments_high_speech / mean_comments * 100 - 100: .2f}%')

Разница Comments между высокой Speechiness и общей:-98.78%


Ответ: Количество комментариев с высоким показателем Speechiness (>0.66) в среднем меньше общего среднего комментариев на 98.78%